### PR TITLE
#3764: Device side opID support

### PR DESCRIPTION
**Pull request by mo-tenstorrent**
### Ticket
[3764](https://github.com/tenstorrent/tt-metal/issues/3764)

### Problem description
Device profiler data did not know which op the running kernels belonged to.

### What's changed

- opID is now updated for cached and uncached FD enqueue programs and for SD LaunchProgram. 
- BRISC FW sets the ID for all riscs ops

With this:

- Device side opID is verified against the expected ID extracted by host
- Tracy device zones include opID
Screenshot 2024-06-28 at 3 11 12 PM

### Checklist
- [ ] Post commit CI passes
- [ ] uBenchmark
- [ ] Device Perf
- [ ] T3K profiler

### CODE REVIEW

File: tt_metal/tools/profiler/kernel_profiler.hpp [hunk 1]
```diff
@@ -167,23 +164,21 @@ namespace kernel_profiler{
             profiler_control_buffer[FLAT_ID] = core_flat_id;
         }
 
-#pragma GCC unroll 65534
         for (int i = GUARANTEED_MARKER_1_H; i < CUSTOM_MARKERS; i ++)
         {
-        //TODO(MO): Clean up magic numbers
+            //TODO(MO): Clean up magic numbers
```
Comment (aliuTT) [inline tt_metal/tools/profiler/kernel_profiler.hpp:169]: where are these numbers getting used? Is there risk that device already holds these values on start, and FW jumps ahead of init?

File: tt_metal/hw/inc/dev_msgs.h [hunk 2]
```diff
@@ -65,6 +65,8 @@ struct launch_msg_t {  // must be cacheline aligned
     volatile uint16_t watcher_kernel_ids[DISPATCH_CLASS_MAX_PROC];
     volatile uint16_t ncrisc_kernel_size16;  // size in 16 byte units
 
+    volatile uint16_t host_assigned_op_id;
```
Comment (aliuTT) [inline tt_metal/hw/inc/dev_msgs.h:68]: @pgkeller was changing `launch_msg_t` in his PR here: https://github.com/tenstorrent/tt-metal/pull/9781/files#diff-d606266ae23fd80f4fa4f80ea161f8b17ff937bbf94cce50cfd731755d0a491b

He suspects that moving `run` field around is causing some issues.

Comment (mo-tenstorrent) [reply tt_metal/hw/inc/dev_msgs.h:68]: Hmm interesting, I did not see any hangs on CI with the addition of the two bytes. I will coordinate before merging this in.

File: tt_metal/tools/profiler/kernel_profiler.hpp (hunk 1 above)
Comment (mo-tenstorrent) [reply tt_metal/tools/profiler/kernel_profiler.hpp:169]: I am referring to the `0x80000000`. They are pretty settled at this point. I will clean them up in another PR

File: tt_metal/impl/program/program.hpp [hunk 3]
```diff
@@ -77,11 +77,13 @@ class Program {
     Program(Program &&other) = default;
     Program& operator=(Program &&other) = default;
 
+    void set_global_id(uint64_t id);
```
Comment (aliuTT) [inline tt_metal/impl/program/program.hpp:80]: Is this only called by ttnn and tteager? So other ops and kernels don't have a global id?

File: tt_metal/tools/profiler/kernel_profiler.hpp (hunk 1 above)
Comment (mo-tenstorrent) [reply tt_metal/tools/profiler/kernel_profiler.hpp:169]: We are in init, so we are flushing out anything that might be there from the previous run.

File: tt_metal/impl/program/program.hpp (hunk 3 above)
Comment (mo-tenstorrent) [reply tt_metal/impl/program/program.hpp:80]: Right, this is for use by higher levels. Making direct tt_metal calls can ignore this and op field will show up as zero ...

Comment (mo-tenstorrent) [reply tt_metal/impl/program/program.hpp:80]: i.e. this is for assigning unique ids for op_report

Comment (aliuTT) [reply tt_metal/impl/program/program.hpp:80]: Where is it initialized to 0?

Comment (mo-tenstorrent) [reply tt_metal/impl/program/program.hpp:80]: Good find, I thought I added it. Will add to initializer list for program

Comment (mo-tenstorrent) [reply tt_metal/impl/program/program.hpp:80]: Added

File: tt_metal/hw/inc/dev_msgs.h [hunk 4]
```diff
@@ -91,7 +93,17 @@ struct launch_msg_t {  // must be cacheline aligned
     volatile uint8_t dispatch_core_x;
     volatile uint8_t dispatch_core_y;
     volatile uint8_t exit_erisc_kernel;
-    volatile uint8_t run;  // must be in last cacheline of this msg
+    volatile uint8_t pad1;
+    volatile uint16_t pad2;
```
Comment (ubcheema) [inline tt_metal/hw/inc/dev_msgs.h:97]: why do we need this 2 byte padding?

Comment (mo-tenstorrent) [reply tt_metal/hw/inc/dev_msgs.h:97]: We need to be 32-bit aligned on this struct, cache aligned.

**Pull request closed without merging**